### PR TITLE
Systeme de vote pour les élections du BDE

### PR DESCRIPTION
**Pull request by MiLk**
01_accueil
02_accueil_noncot
03_after_vote

### CODE REVIEW

**Comment by mattgu74** [comment]
Tu anonymyse bien les vites dans la base ?

J'ai cru le voir à partir des schéma de la table et notamment cette ligne :

count: { type: integer, notnull: true, default: 0 }

Tu confirmes ? Et tu ne généres pas des log quelconques qui pourrait
permettre de dire l'user X à voté la liste Y ?
Si oui c'est parfait et ça m'a l'air d'être le cas ;) Et j'aime beaucoup
les screenshots, simple et efficace pour faire accepter une pull request ;)

2013/2/26 MiLk notifications@github.com

> [image: 01_accueil]https://f.cloud.github.com/assets/29782/195679/924172fa-7ffb-11e2-9e2c-b4d740c05355.png
> [image: 02_accueil_noncot]https://f.cloud.github.com/assets/29782/195680/92428e88-7ffb-11e2-9251-159a856e4964.png
> 
> ## [image: 03_after_vote]https://f.cloud.github.com/assets/29782/195681/9257304a-7ffb-11e2-8b6d-2aa04eed5673.png
> 
> You can merge this Pull Request by running
> 
>   git pull https://github.com/MiLk/portail vote
> 
> Or view, comment on, or merge it at:
> 
>   https://github.com/simde-utc/portail/pull/22
> Commit Summary
> - Systeme de vote pour les élections du BDE
> - Confirmation javascript
> 
> File Changes
> - _M_ apps/backend/config/routing.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-0(9)
> - _A_ apps/backend/modules/vote_liste/actions/actions.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-1(16)
> - _A_ apps/backend/modules/vote_liste/config/generator.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-2(21)
> - _A_
>   apps/backend/modules/vote_liste/lib/vote_listeGeneratorConfiguration.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-3(13)
> - _A_
>   apps/backend/modules/vote_liste/lib/vote_listeGeneratorHelper.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-4(13)
> - _M_ apps/backend/templates/layout.phphttps://github.com/simde-utc/portail/pull/22/files#diff-5(2)
> - _A_ apps/vote/config/app.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-6(5)
> - _A_ apps/vote/config/cache.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-7(7)
> - _A_ apps/vote/config/factories.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-8(42)
> - _A_ apps/vote/config/filters.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-9(10)
> - _A_ apps/vote/config/routing.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-10(21)
> - _A_ apps/vote/config/security.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-11(5)
> - _A_ apps/vote/config/settings.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-12(40)
> - _A_ apps/vote/config/view.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-13(20)
> - _A_ apps/vote/config/voteConfiguration.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-14(8)
> - _A_ apps/vote/lib/myUser.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-15(38)
> - _A_ apps/vote/modules/cas/actions/actions.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-16(36)
> - _A_ apps/vote/modules/cas/config/security.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-17(5)
> - _A_ apps/vote/modules/vote/actions/actions.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-18(65)
> - _A_ apps/vote/modules/vote/config/security.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-19(2)
> - _A_ apps/vote/modules/vote/templates/indexSuccess.phphttps://github.com/simde-utc/portail/pull/22/files#diff-20(27)
> - _A_ apps/vote/templates/layout.phphttps://github.com/simde-utc/portail/pull/22/files#diff-21(94)
> - _M_ config/doctrine/schema.ymlhttps://github.com/simde-utc/portail/pull/22/files#diff-22(26)
> - _A_ lib/filter/doctrine/VoteFormFilter.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-23(16)
> - _A_ lib/filter/doctrine/VoteListeFormFilter.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-24(16)
> - _A_ lib/form/doctrine/VoteForm.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-25(16)
> - _A_ lib/form/doctrine/VoteListeForm.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-26(17)
> - _A_ lib/migration/doctrine/1361869421_version22.phphttps://github.com/simde-utc/portail/pull/22/files#diff-27(127)
> - _A_ lib/migration/doctrine/1361869422_version23.phphttps://github.com/simde-utc/portail/pull/22/files#diff-28(75)
> - _A_ lib/model/doctrine/Vote.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-29(15)
> - _A_ lib/model/doctrine/VoteListe.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-30(15)
> - _A_ lib/model/doctrine/VoteListeTable.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-31(26)
> - _A_ lib/model/doctrine/VoteTable.class.phphttps://github.com/simde-utc/portail/pull/22/files#diff-32(27)
> - _A_ test/functional/backend/vote_listeActionsTest.phphttps://github.com/simde-utc/portail/pull/22/files#diff-33(19)
> - _A_ test/functional/vote/voteActionsTest.phphttps://github.com/simde-utc/portail/pull/22/files#diff-34(19)
> - _A_ web/css/votes.csshttps://github.com/simde-utc/portail/pull/22/files#diff-35(20)
> - _A_ web/vote.phphttps://github.com/simde-utc/portail/pull/22/files#diff-36(7)
> - _A_ web/vote_dev.phphttps://github.com/simde-utc/portail/pull/22/files#diff-37(13)
> 
> Patch Links:
> - https://github.com/simde-utc/portail/pull/22.patch
> - https://github.com/simde-utc/portail/pull/22.diff

**Comment by MiLk** [comment]
Pour chaque vote, j'enregistre l'user, le semestre (pour pouvoir réutiliser le même système par la suite sans réinitialiser), et l'heure du vote. Je ne l'associe pas à la liste.
J'incrémente séparément le nombre de vote sur la liste choisie.

**Comment by apuyou** [comment]
Les screens c'est la vie, mais bon faut regarder le code quand même hein ^^

Je pourrais jeter un coup d'œil ce soir. 

Arthur Puyou

**Comment by MiLk** [comment]
Il faut ajouter le vote blanc.

File: apps/vote/lib/myUser.class.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  {
+    sfCAS::initPhpCAS();
+    phpCAS::forceAuthentication();
+    $username = phpCAS::getUser();
+
+    $data = sfGuardUserTable::getInstance()->findOneBy('username', $username);
+    if($data && $data->getIsActive())
+      $this->signin($data, true);
+    else
+      die('Unauthorized.');
```
Comment (apuyou) [inline apps/vote/lib/myUser.class.php]: J'avais modifié ce fichier dans l'appli frontend pour éviter le die qui est atroce pour l'utilisateur : https://github.com/simde-utc/portail/blob/master/apps/frontend/lib/myUser.class.php

AMHA, si on n'appelle pas signin, on se retrouve juste pas connecté après, donc pas besoin de stopper l'exécution de la page.

Il n'y a pas non plus la fonction register. Vu que `login()` cherche dans la `sfGuardUserTable`, je comprends qu'il faut que l'utilisateur ait **déjà** un compte sur le portail pour pouvoir aller voter ? Si oui, c'est bof.

Si on peut éviter de lui créer un compte portail juste pour ça, c'est bien, mais sinon il faut le faire ici.

File: apps/vote/modules/vote/actions/actions.class.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    $this->listes = VoteListeTable::getInstance()->getCurrentLists()->execute();
+    $this->is_cotisant = $this->isCotisant();
+    $this->has_voted = VoteTable::getInstance()->getVoteForUserAndSemestre($this->getUser()->getGuardUser()->getPrimaryKey(),sfConfig::get('app_portail_current_semestre'))->fetchOne();
+  }
+
+  public function executeVote(sfWebRequest $request) {
+    $liste = $this->getRoute()->getObject();
+    if(!$liste) {
+      $this->getUser()->setFlash('error','Cette liste n\'existe pas.');
+    } else
```
Comment (apuyou) [inline apps/vote/modules/vote/actions/actions.class.php]: Sérieux faut mettre des accolades ! :confused:

Comment (apuyou) [inline apps/vote/modules/vote/actions/actions.class.php]: Oh c'est un `else if` ! :confused: :confused:

File: apps/vote/modules/vote/templates/indexSuccess.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<h1>Élections du nouveau BDE</h1>
+<h2>1. Je consulte les programmes</h2>
+<?php foreach($listes as $liste): ?>
+  <div class="well liste">
+    <h3><?php echo $liste->getNom() ?></h3>
+    <h4>Programme:</h4>
```
Comment (apuyou) [inline apps/vote/modules/vote/templates/indexSuccess.php]: Virer les `:` dans le titre

File: apps/vote/modules/vote/templates/indexSuccess.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    <p>
+      <?php echo $liste->getDescription(ESC_RAW) ?>
+    </p>
+    <h4>Membres:</h4>
```
Comment (apuyou) [inline apps/vote/modules/vote/templates/indexSuccess.php]: Virer les `:` dans le titre

File: apps/vote/templates/layout.php [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+          <a class="brand" href="<?php echo url_for('homepage') ?>">Élections BDE</a>
+          <div class="btn-group pull-right">
+            <?php if(!$sf_user->isAuthenticated()): ?>
+              <a class="btn dropdown-toggle" data-toggle="dropdown" href="#drop-connexion">
+                <i class="icon-user"></i> Connexion
+                <b class="caret"></b>
+              </a>
+              <ul class="dropdown-menu pull-right">
+                <li><a href="<?php echo url_for('cas') ?>">Je suis étudiant</a></li>
+                <li><a href="<?php echo url_for('sf_guard_signin') ?>">Extérieur</a></li>
```
Comment (apuyou) [inline apps/vote/templates/layout.php]: À virer

File: apps/vote/lib/myUser.class.php (hunk 1 above)
Comment (MiLk) [inline apps/vote/lib/myUser.class.php]: L'utilisation du compte du portail permet d'empêcher le double vote plus simplement.
Si on retire complètement l'utilisation de la DB utilisateur, il faudra aussi gérer la session manuellement.

J'ai recup le code du backend, donc effectivemment il faut faire des changements que l'on décide d'utiliser les comptes du portail ou non.

File: apps/vote/modules/vote/actions/actions.class.php (hunk 2 above)
Comment (MiLk) [inline apps/vote/modules/vote/actions/actions.class.php]: Ce sont des jolis else if multilignes !
Mais bon on peut remettre sur une ligne si tu préfères.

File: apps/vote/lib/myUser.class.php (hunk 1 above)
Comment (apuyou) [inline apps/vote/lib/myUser.class.php]: OK bah fais des comptes portail, c'est pas grave :)

**Comment by apuyou** [comment]
Deux remarques pour le backend :
- masquer les created_at/updated_at
- masquer le count (ça craint de pouvoir le modifier aussi facilement)

J'ai mis un commentaire sur la ligne à changer.

Pour le reste, c'est bon pour moi :smiley:

File: lib/form/doctrine/VoteListeForm.class.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+ * @package    simde
+ * @subpackage form
+ * @author     Your name here
+ * @version    SVN: $Id: sfDoctrineFormTemplate.php 23810 2009-11-12 11:07:44Z Kris.Wallsmith $
+ */
+class VoteListeForm extends BaseVoteListeForm
+{
+  public function configure()
+  {
+    unset($this['created_at'], $this['updated_at']);
```
Comment (apuyou) [inline lib/form/doctrine/VoteListeForm.class.php]: Pourquoi ça ne marche pas les trucs des dates ? :(

Ajouter `$this['count']`.

**Comment by MiLk** [comment]
Pour created_at et updated_at c'est que au niveau des filters non ?
J'ai désactivé dans le form et les filters.
J'ai viré le count du form.
Pense à bien vider ton cache.

**Comment by apuyou** [comment]
C'est parfait, j'ai testé ça marche nickel ! :smiley: 

Est-ce que tu pull ou je le fais ?
Il faut aussi ajouter la config dans nginx en https pour avoir une URL jolie (et gérer le election.php de toute façon).

**Pull request merged**